Adding support for smaller screens (ie. mobile support) so that nav menu toggles

RichieEscarez opened the pull request:
- Restructured layout to use bootstrap columns so that nav menu and doc content dynamically change size depending on the size of the window
- Nav menu automatically toggles closed on screens smaller than 768px
- Added a button and support to toggle open/closed the nav menu 
- Dynamically resize search field to accommodate window size and added nav menu toggle button
- Nav menu auto scrolls to topic when toggled open

Preview these changes in my fork @ http://richieescarez.github.io/kubernetes/v1.0

Fixes #12157

RichieEscarez committed:
Adding support for smaller screens (ie. mobile support) so that nav menu toggles
Fixes #12157

thockin commented:
When I load your fork on chrome/linux desktop I have a huge margin left of
the left menu.

When I load this on chrome/Android I get a top-bar (teal with up arrow that
does nothing), a huge left and right margin, a folded-up arrow which
expands to the left menu, and a tiny centered form of the title slide, just
showing "Docume"

So I'm gonna say this needs a BIT more testing

On Wed, Sep 23, 2015 at 10:37 AM, RichieEscarez <notifications@github.com>
wrote:

>
>    - Restructured layout to use bootstrap columns so that nav menu and
>    doc content dynamically change size depending on the size of the window
>    - Nav menu automatically toggles closed on screens smaller than 768px
>    - Added a button and support to toggle open/closed the nav menu
>    - Dynamically resize search field to accommodate window size and added
>    nav menu toggle button
>    - Nav menu auto scrolls to topic when toggled open
>
> Preview these changes in my fork @
> http://richieescarez.github.io/kubernetes/v1.0
>
> Fixes #12157 <https://github.com/kubernetes/kubernetes/issues/12157>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/kubernetes/kubernetes/pull/14436
> Commit Summary
>
>    - Adding support for smaller screens (ie. mobile support) so that nav
>    menu toggles
>
> File Changes
>
>    - *M* _layouts/docwithnav.html
>    <https://github.com/kubernetes/kubernetes/pull/14436/files#diff-0>
>    (111)
>    - *M* _sass/_documents.scss
>    <https://github.com/kubernetes/kubernetes/pull/14436/files#diff-1>
>    (36)
>    - *M* _sass/_md.scss
>    <https://github.com/kubernetes/kubernetes/pull/14436/files#diff-2>
>    (13)
>    - *M* _sass/_navigation.scss
>    <https://github.com/kubernetes/kubernetes/pull/14436/files#diff-3>
>    (18)
>    - *M* _sass/_sm.scss
>    <https://github.com/kubernetes/kubernetes/pull/14436/files#diff-4>
>    (14)
>    - *M* _sass/_xs.scss
>    <https://github.com/kubernetes/kubernetes/pull/14436/files#diff-5>
>    (14)
>    - *A* img/mobile/chevron-left.png
>    <https://github.com/kubernetes/kubernetes/pull/14436/files#diff-6> (0)
>    - *A* img/mobile/chevron-right.png
>    <https://github.com/kubernetes/kubernetes/pull/14436/files#diff-7> (0)
>    - *M* v1.0/index.md
>    <https://github.com/kubernetes/kubernetes/pull/14436/files#diff-8> (2)
>
> Patch Links:
>
>    - https://github.com/kubernetes/kubernetes/pull/14436.patch
>    - https://github.com/kubernetes/kubernetes/pull/14436.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14436>.
>

RichieEscarez commented:
A whole can of worms got opened with adding this mobile support (Ive been fixing these types of little things for a day now).
Questions/responses:
 - Should the default behavior be to keep the nav menu open (instead of closing it and showing the arrow)?
 - I know how to fix the "back to top" arrow (looks like I accidentally overwrote that click behavior).
 - I also noticed the "Documentation" banner cutoff and planned to take that on in a separate PR. It currently reuses the code used in the other pages but might need a specific version for the Documentation section. (Although I could try making it dynamically detect page size so it automatically reduces font size).
 - For the linux/chome view (desktop view). The way the bootstrap dynamic columns work is that they resize proportionally to three differently set ratios depending on ranges of window sizes. So within a given range, the left margin might be larger when window size is near the top of that range. Also, the doc content view was limited to 900px at max (which i left in). Since there is a limit for the doc content (900px), i also set the max width of the nav menu to 255px to stop it from growing larger than the width of the content. Question: Should i remove the limit of the doc content (900px) and allow both the nav menu and content proportionally file the screen (3/9 ratio)?

RichieEscarez commented:
The up arrow i the teal header seems to be working fine for me. That has been there (i didnt add it) and is meant to act as a "back to top" button. Is it not working for you? Or are you commenting on how it provides not function when already scrolled to the top of the page? Im also seeing this behavior out on kubernetes.io now.

Question: Should I try to hide it until the page actually scrolls?

RichieEscarez commented:
@thockin I tweaked a bunch of margin and padding settings and also found and used another bootstrap class, which dynamically sizes content per window size (not per window size range, as it was before). Another difference is now all content is left justified (ive not been able to figure out how to change that, and might not be able to). PTAL and let me know your feedback for the current behavior.

thockin commented:
Desktop:  The title slide now underlays the custom-search bar.

Mobile: title slide is not centered, still not wide enough for the word
"Documentation", has a large right margin, and when I expand the menu, it
all shifts.  Again, I don't know how to do this right.  I'm just a poor
backend guy.

On Wed, Sep 23, 2015 at 3:36 PM, RichieEscarez <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> I tweaked a bunch of margin and
> padding settings and also found and used another bootstrap class, which
> dynamically sizes content per window size (not per window size range, as it
> was before). Another difference is now all content is left justified (ive
> not been able to figure out how to change that, and might not be able to).
> PTAL and let me know your feedback for the current behavior.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14436#issuecomment-142749781>
> .
>

thockin commented:
Also see left-justifeed, though that seems OK?  Not sure.

On Wed, Sep 23, 2015 at 3:43 PM, Tim Hockin <thockin@google.com> wrote:

> Desktop:  The title slide now underlays the custom-search bar.
>
> Mobile: title slide is not centered, still not wide enough for the word
> "Documentation", has a large right margin, and when I expand the menu, it
> all shifts.  Again, I don't know how to do this right.  I'm just a poor
> backend guy.
>
> On Wed, Sep 23, 2015 at 3:36 PM, RichieEscarez <notifications@github.com>
> wrote:
>
>> @thockin <https://github.com/thockin> I tweaked a bunch of margin and
>> padding settings and also found and used another bootstrap class, which
>> dynamically sizes content per window size (not per window size range, as it
>> was before). Another difference is now all content is left justified (ive
>> not been able to figure out how to change that, and might not be able to).
>> PTAL and let me know your feedback for the current behavior.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/14436#issuecomment-142749781>
>> .
>>
>
>

RichieEscarez commented:
@thockin, ok thanks.
 * Ill work on the "Documentation" banner so that it shrinks down to fix (and also fix the overlapping with the search field).
 * Ill also take a different approach to how the nav menu displays in these smaller views. Instead of sharing the view, ill change it so that the nav menu overlays the help topic (the resizing and sharing of the view is whats causing the "shifting"). In addition, ill add that any click on the nav menu (when in "mobile view") closes it and displays the topic.

timstclair commented:
Have you considered moving the nav to a subsection under Documentation in the top-level page menu (on mobile only)? e.g. something like this (obviously with some tweaking, no scrollbars, etc.):


![oj549puizkq](https://cloud.githubusercontent.com/assets/3272040/10112681/ff26621e-6391-11e5-9282-917afcb1edd1.png)

RichieEscarez commented:
@timstclair , only briefly did I consider it because I decided that it would cost a lot more to make happen. All the javascript we use for loading content and animating the nav menu would have to be modified so given my list of priorities, i wanted to try and avoid changing more that hiding/showing divs. Maybe open a enhancement request to sit in the backlog for now?

brendandburns commented:
@timstclair I'm delegating to you for LGTM.  Let me know when you're happy and we'll get this in.

--brendan

timstclair commented on _layouts/docwithnav.html:
```diff
@@ -65,49 +78,61 @@
-					animatenav(findcurnavitem());//is an in-page link
-				} else {
-					animatenav(clickedon);//is a nav menu link
+	            return false;
```
(nit) indentation

RichieEscarez replied:
Looks right in the file but was a mix of spaces and tabs so ive changed them all to tabs to match how the } above it is indented (maybe that makes a difference in how it displays here in the compare view).

RichieEscarez replied:
fixed

timstclair commented on _sass/_md.scss:
```diff
@@ -1,4 +1,20 @@
+	.searchfield {
+		top: 82px;
+        width: 85vw;
+        max-width: 1100px;
```
Where'd 1100px come from? Why not either full width or line up with the right edge of the content?

RichieEscarez replied:
Sorry, this was left over from troubleshooting. I meant to set this to max width of nav menu plus max width of the content body. Both the nav menu and content body have set width (255 + 900px), so this should be 1155px. My intention was to avoid having the search box span the whole screen and instead only float over width of the nav menu and content body. I just thought it was odd to have the [Search] button so far off to the top left of the screen on its own.

Im fine with backing out my changes and letting it span the whole screen because I think a better solution might be to hide the search box into a icon (magnifying glass) in the banner so that we can gain some real estate back (without loosing its constant presence). Any thoughts here (this would be a backlog item)?

timstclair commented on _sass/_md.scss:
```diff
@@ -1,4 +1,20 @@
 @media #{$md} {
+	.searchfield {
+		top: 82px;
+        width: 85vw;
```
I think it's a bad idea to mix fixed pixel sizes with relative dimensions.

RichieEscarez replied:
I think i can remove the "vw" setting now. I needed this before I found the class="container-float" option (the search field was not resizing to match the screen width and extended off the screen).

RichieEscarez commented:
FYI: Im starting on these two changes (as mentioned above). Let me know otherwise.

 * Ill work on the "Documentation" banner so that it shrinks down to fix (and also fix the overlapping with the search field).
 * Ill also take a different approach to how the nav menu displays in these smaller views. Instead of sharing the view, ill change it so that the nav menu overlays the help topic (the resizing and sharing of the view is whats causing the "shifting"). In addition, ill add that any click on the nav menu (when in "mobile view") closes it and displays the topic.

RichieEscarez commented:
@timstclair PTAL, ive added in the following:
* removed sharing view (nav menu now overlays content when toggled in small views) 
* automatically close the nav menu on click 
* integrated in the recent searchbox improvements (#14591)
* improved bootstrap layout configuration (improved behavior on viewport resize)
* automatically reduce text size in the documentation banner

RichieEscarez commented:
I changed the toggle chevron color to highlight it too - let me know if black is better.

Also, about the nav menu auto close, one thought from testing is that I could delay closing the nav menu for any nav menu containers? 

I think most if not all nav menu containers are actual topics (ie it opens a help topic) but by delaying the nav menu close, (pro) user will see all the sub topics that the container includes? But, (con) they will need to manually close the nav menu in order to see the help topic that containers open.

Any suggestions here?

RichieEscarez commented:
Just noticing in my fork that the chevron is not toggling correctly. Ill troubleshoot what went wrong with the cherry-pick into my gh-pages branch later.

timstclair commented on _layouts/docwithnav.html:
```diff
@@ -1,15 +1,31 @@
+			{% include documentation.html %}
+		</div>
+	</div>
+	<div class="gcse-wrapper">
```
I would keep this as a top-level container. It displays floating over everything else, so there's no reason to nest it

RichieEscarez replied:
Ok, moved it outside the <div class="container-fluid">...</div>

timstclair commented on _sass/_navigation.scss:
```diff
@@ -91,7 +97,22 @@ ul, ul ul {
-#gcse-wrapper {
-  z-index: 1;
-  position: fixed;
+#gcse-searchbox {
```
You lost the padding on the search box. I think it looks cramped without it.

RichieEscarez replied:
Adding in: ```padding: 13px;``` and adjusted the views

timstclair commented:
RE: nav menu close delay. My suggestion is if you tap a nav item which expands, the first tap should expand the menu & keep the nav open. The second tap should close the nav.

brendandburns commented:
needs rebase...

RichieEscarez commented:
(1) added in searchbox padding
(2) adjusted searchbox position across all views (created and used global variables)
(3) adjusted nav menu toggle button (and fixed chevron toggle)
(4) fixed generated TOCs so that they only load the TOC into the page (avoid whole page refresh) because it was toggleing the nav menu on container clicks
(5) added behavior to keep open containers so that when they are clicked, they display sub-topics instead of closing (clicking a non container topic in the nav menu will close the nav menu)
(6) also added the ability to toggle open/closed containers when clicked twice (found it annoying to have to scroll long open menus when in a small res. display (ie. phone))

RichieEscarez commented:
will now rebase, resolve conflicts, and then push to my fork

RichieEscarez commented:
@timstclair PTAL http://richieescarez.github.io/kubernetes/v1.0

timstclair commented:
I think this looks good (LGTM).

The 3 features I would add later are:

1. Tapping outside the nav panel should close it
2. Some visual separation of the nav panel (e.g. a [box-shadow](http://www.w3schools.com/cssref/css3_pr_box-shadow.asp))
3. Move the search bar up into the top nav bar (blue)

@brendandburns did you have anything to add?

RichieEscarez commented:
1. and 2. were quick and easy so I just added them.

I added code for 1. here: https://github.com/kubernetes/kubernetes/pull/14436/files#diff-1a1c9c2581e96ce455568c50db4124d4R86

And inserted ```box-shadow: 2px 2px 5px #888888;``` in the _xs.scss and _sm.scss files for 2.

timstclair commented:
Nice. Still LGTM.

RichieEscarez commented:
Thanks @timstclair! Im adding the LGTM label. @brendandburns if you have something more, just let me know.

RichieEscarez commented:
Issue for improvements to the search box: #15092
